### PR TITLE
Enhance CA handling and error messaging for improved configuration usability

### PR DESCRIPTION
**Pull request by GabrielEValenzuela**
## Objective
This PR enhances the system's ability to handle multiple Certificate Authorities (CAs) by modifying the logic to read multiple CA certificates from the configuration and concatenate them into a single file. Additionally, we have improved error messaging to enhance usability, aligning with the Ghezzi quality software characteristics.

## Description
We implemented a new logic that allows the system to read multiple CA entries from the configuration and concatenate them into a single CA bundle file. This streamlined approach simplifies the setup process for scenarios requiring multiple CAs, this approach has been validated as 

-  #24551

We also recognized the need to improve the clarity and helpfulness of error messages related to CA configuration. Following the Ghezzi quality software characteristic of usability, we refined our error messaging to be more descriptive, guiding users more effectively through resolving configuration issues.

### Static Analysis
N/A

#### Unit Testing
N/A

#### Integration Testing
- Integration testing was conducted to ensure that the system correctly handles the new CA bundle during runtime, especially in environments requiring secure communications authenticated by multiple CAs.
- **Testing Steps**:
  1. Configure the system with multiple valid and invalid CA entries. Follow the first step of #24551
  2. Verify that the system concatenates valid CAs correctly and utilizes the resulting CA bundle for secure connections. Use the `cat` command for example and `stat` to check file permissions.
  3. Run manager, and indexer connection should be successful.

<details><summary>Screenshots</summary>

image

Pasted image

image

image

</details>

## Additional Notes
- The path proposal can change according with reviewers.

### CODE REVIEW

File: src/shared_modules/indexer_connector/src/indexerConnector.cpp [hunk 1]
```diff
@@ -58,7 +58,50 @@ static void initConfiguration(SecureCommunication& secureCommunication, const nl
...
+            if (filePaths.size() > 1)
+            {
+                std::string caRootCertificateContentMerged;
+
+                for (const auto& filePath : filePaths)
+                {
+                    if (!std::filesystem::exists(filePath))
+                    {
+                        throw std::runtime_error("The CA root certificate file: " + filePath +
+                                                 " does not exist. Did you check the path?");
```
Comment (sebasfalcone) [inline src/shared_modules/indexer_connector/src/indexerConnector.cpp:73]:
```suggestion
                        throw std::runtime_error("The CA root certificate file: '" + filePath +
                                                 "' does not exist.");
```

File: src/shared_modules/indexer_connector/src/indexerConnector.cpp [hunk 2]
```diff
@@ -58,7 +58,50 @@ static void initConfiguration(SecureCommunication& secureCommunication, const nl
...
+                    }
+
+                    std::ifstream file(filePath);
+                    if (!file.is_open())
+                    {
+                        throw std::runtime_error("Could not open CA root certificate file: " + filePath +
+                                                 ". Did you check the permissions?");
```
Comment (sebasfalcone) [inline src/shared_modules/indexer_connector/src/indexerConnector.cpp:80]: Issue could be due to other factor
```suggestion
                        throw std::runtime_error("Could not open CA root certificate file: '" + filePath +
                                                 "'.");
```

File: src/shared_modules/indexer_connector/src/indexerConnector.cpp [hunk 3]
```diff
@@ -58,7 +58,50 @@ static void initConfiguration(SecureCommunication& secureCommunication, const nl
...
+                std::ofstream outputFile(caRootCertificate);
+                if (outputFile.is_open())
+                {
+                    outputFile << caRootCertificateContentMerged;
+                    outputFile.close();
+                    logDebug2(IC_NAME, "All CA files merged into '%s' successfully.", caRootCertificate.c_str());
+                }
+                else
+                {
+                    throw std::runtime_error("Could not write the CA root merged file");
```
Comment (sebasfalcone) [inline src/shared_modules/indexer_connector/src/indexerConnector.cpp:98]:
```suggestion
                    throw std::runtime_error("Could not write the CA root merged file.");
```

File: src/shared_modules/indexer_connector/src/indexerConnector.cpp [hunk 4]
```diff
@@ -58,7 +58,50 @@ static void initConfiguration(SecureCommunication& secureCommunication, const nl
...
+                        throw std::runtime_error("Could not open CA root certificate file: " + filePath +
+                                                 ". Did you check the permissions?");
+                    }
+
+                    caRootCertificateContentMerged.append((std::istreambuf_iterator<char>(file)),
+                                                          std::istreambuf_iterator<char>());
+                }
+
+                // Once all the files are read, we write the content into final file.
+                caRootCertificate = "/etc/wazuh-indexer/certs/root-ca-merged.pem";
```
Comment (sebasfalcone) [inline src/shared_modules/indexer_connector/src/indexerConnector.cpp:88]: We should use the installation path for this (it has the right permissions)

```suggestion
                caRootCertificate = "/etc/wazuh/etc/root-ca-merged.pem";
```

**Review by sebasfalcone — CHANGES_REQUESTED** [review]

File: src/shared_modules/indexer_connector/src/indexerConnector.cpp [hunk 5]
```diff
@@ -58,7 +62,62 @@ static void initConfiguration(SecureCommunication& secureCommunication, const nl
...
+                    {
+                        throw std::runtime_error("Could not open CA root certificate file: '" + filePath + "'.");
+                    }
+
+                    caRootCertificateContentMerged.append((std::istreambuf_iterator<char>(file)),
+                                                          std::istreambuf_iterator<char>());
+                }
+
+                // Once all the files are read, we write the content into final file.
+                caRootCertificate = "/var/ossec/var/certs/root-ca-merged.pem";
```
Comment (sebasfalcone) [inline src/shared_modules/indexer_connector/src/indexerConnector.cpp:90]: As @MiguelazoDS stated, lets check if we can make this path relative

Comment (GabrielEValenzuela) [reply src/shared_modules/indexer_connector/src/indexerConnector.cpp:90]: If the folder exists the result is successfully

```console
2024/07/18 13:16:58 indexer-connector[8114] indexerConnector.cpp:115 at initConfiguration(): DEBUG: All CA files merged into 'var/certs/root-ca-merged.pem' successfully.
```

Comment (GabrielEValenzuela) [reply src/shared_modules/indexer_connector/src/indexerConnector.cpp]: Changes done!

**Review by sebasfalcone — APPROVED** [review]
LGTM! Great testing @GabrielEValenzuela

**Review by Dwordcito — CHANGES_REQUESTED** [review]

File: src/init/inst-functions.sh [hunk 6]
```diff
@@ -1052,6 +1052,7 @@ InstallLocal()
     ${INSTALL} -d -m 0770 -o ${WAZUH_USER} -g ${WAZUH_GROUP} ${INSTALLDIR}/var/multigroups
     ${INSTALL} -d -m 0770 -o root -g ${WAZUH_GROUP} ${INSTALLDIR}/var/db
     ${INSTALL} -d -m 0770 -o root -g ${WAZUH_GROUP} ${INSTALLDIR}/var/download
+    ${INSTALL} -d -m 0770 -o root -g ${WAZUH_GROUP} ${INSTALLDIR}/var/certs
```
[the remark this answers is not in the document]
Comment (GabrielEValenzuela) [reply src/init/inst-functions.sh]: Remove!

File: src/shared_modules/indexer_connector/src/indexerConnector.cpp [hunk 7]
```diff
@@ -58,7 +63,74 @@ static void initConfiguration(SecureCommunication& secureCommunication, const nl
...
+                    {
+                        throw std::runtime_error("Could not open CA root certificate file: '" + filePath + "'.");
+                    }
+
+                    caRootCertificateContentMerged.append((std::istreambuf_iterator<char>(file)),
+                                                          std::istreambuf_iterator<char>());
+                }
+
+                // Once all the files are read, we write the content into final file.
+                caRootCertificate = "var/certs/root-ca-merged.pem";
```
[the remark this answers is not in the document]
Comment (GabrielEValenzuela) [reply src/shared_modules/indexer_connector/src/indexerConnector.cpp]: Done!

File: src/shared_modules/indexer_connector/src/indexerConnector.cpp [hunk 8]
```diff
@@ -58,7 +63,74 @@ static void initConfiguration(SecureCommunication& secureCommunication, const nl
         if (config.at("ssl").contains("certificate_authorities") &&
             !config.at("ssl").at("certificate_authorities").empty())
         {
-            caRootCertificate = config.at("ssl").at("certificate_authorities").front().get_ref<const std::string&>();
+            std::vector<std::string> filePaths =
+                config.at("ssl").at("certificate_authorities").get<std::vector<std::string>>();
+
+            if (filePaths.size() > 1)
+            {
+                std::string caRootCertificateContentMerged;
```
[the remark this answers is not in the document]
Comment (GabrielEValenzuela) [reply src/shared_modules/indexer_connector/src/indexerConnector.cpp]: Done!

File: src/shared_modules/indexer_connector/src/indexerConnector.cpp [hunk 9]
```diff
@@ -58,7 +63,74 @@ static void initConfiguration(SecureCommunication& secureCommunication, const nl
...
+                    logDebug2(IC_NAME, "All CA files merged into '%s' successfully.", caRootCertificate.c_str());
+                }
+                else
+                {
+                    throw std::runtime_error("Could not write the CA root merged file");
+                }
+            }
+            else
+            {
+                caRootCertificate = filePaths.at(0);
```
[the remark this answers is not in the document]
Comment (GabrielEValenzuela) [reply src/shared_modules/indexer_connector/src/indexerConnector.cpp]: Done!

File: src/shared_modules/indexer_connector/src/indexerConnector.cpp [hunk 10]
```diff
@@ -58,7 +63,74 @@ static void initConfiguration(SecureCommunication& secureCommunication, const nl
...
+                // Create the directory if it does not exist.
+                std::filesystem::path dirPath = std::filesystem::path(caRootCertificate).parent_path();
+
+                if (!std::filesystem::exists(dirPath))
+                {
+                    if (!std::filesystem::create_directories(dirPath))
+                    {
+                        throw std::runtime_error("Could not create the directory for the CA root merged file");
+                    }
+                }
```
[the remark this answers is not in the document]
Comment (GabrielEValenzuela) [reply src/shared_modules/indexer_connector/src/indexerConnector.cpp]: Done!

File: src/shared_modules/indexer_connector/src/indexerConnector.cpp [hunk 11]
```diff
@@ -45,6 +51,59 @@ constexpr auto SYNC_QUEUE_LIMIT = 4096;
...
+            throw std::runtime_error("Could not open CA root certificate file: '" + filePath + "'.");
+        }
+
+        caRootCertificateContentMerged.append((std::istreambuf_iterator<char>(file)), std::istreambuf_iterator<char>());
+    }
+
+    caRootCertificate = DEFAULT_PATH;
+
+    std::filesystem::path dirPath = std::filesystem::path(caRootCertificate).parent_path();
+    if (!std::filesystem::exists(dirPath) && !std::filesystem::create_directories(dirPath))
```
Comment (Dwordcito) [inline src/shared_modules/indexer_connector/src/indexerConnector.cpp:77]:
```suggestion
    if (std::filesystem::path dirPath = std::filesystem::path(caRootCertificate).parent_path();
    !std::filesystem::exists(dirPath) && !std::filesystem::create_directories(dirPath))
```

**Review by Dwordcito — APPROVED** [review]

**Pull request merged**